### ISSUE TITLE
Fully programable depth compare and blending

### ISSUE THREAD

**Issue by standard-two-simplex** [gonetz/GLideN64#2599]
I noticed that extension [GL_ARB_TEXTURE_BARRIER](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_texture_barrier.txt), included in the [OpenGL 4.5 Core Profile](https://www.khronos.org/registry/OpenGL/specs/gl/glspec45.core.pdf) (section 9.3), allows reading the same framebuffer object used as color attachment under certain conditions. 

Quote:

> Specifically, the values of rendered fragments are undefined if any shader stage
fetches texels and the same texels are written via fragment shader outputs, even
if the reads and writes are not in the same draw call, unless any of the following
exceptions apply:
• ...
• There is only a single read and write of each texel, and the read is in
the fragment shader invocation that writes the same texel (e.g. using
texelFetch2D(sampler, ivec2(gl_FragCoord.xy), 0);).
• ...

In other words, it is possible to do N64 style depth compare and blending in a similar way as the framebuffer fetch extensions do it for GLES. It has the upside of a wide support on modern GL devices, not needing fragment interlock extensions and being able to do the blending equation without mixing the fragment and OpenGL's built in blender.

The extension seems to be used in current master, but I don't know what for.

I won't be working on this anytime soon, so I thought I would share the knowledge, in case anyone wants to take the task.

**Comment by gonetz** [comment]
I'll try to take a look, thanks!

**Comment by gonetz** [comment]
I've read about glTextureBarrier:
[Texture_barrier](https://www.khronos.org/opengl/wiki/Memory_Model#Texture_barrier)
As I understood, it is not analogue of framebuffer fetch extensions for GLES.
It rather analogue of glMemoryBarrier for framebuffers.
Cite:
```
You just issue a barrier anytime you draw something that may have overlapped with some other blended object you drew. Thus, the above becomes:
Render an object that reads from image, blends in the shader, and writes to image.
Call glTextureBarrier, if the next object overlaps with the first.
Render an object that reads from image, blends in the shader, and writes to image.
Call glTextureBarrier, if the next object overlaps with the first.
Repeat as needed for each blended object.
```

It is the problem. Why "compatible" N64 depth compare mode is slow?
Because we call draw for each triangle, and call glMemoryBarrier for synchronization:
```
for (GLint i = 0; i < GLint(_params.verticesCount); i += 3) {
	glMemoryBarrier(GL_SHADER_IMAGE_ACCESS_BARRIER_BIT);
	glDrawArrays(GLenum(_params.mode), vboStartPos + i, 3);
}
```
It is necessary because triangles usually overlaps.
As I understand, we need to do the same with glTextureBarrier.
The only advantage is that we can use regular depth buffer instead of an image texture.
Please correct me if I'm wrong.

**Comment by standard-two-simplex** [comment]
As I understand, the extension has several functionalities.

> You just issue a barrier anytime you draw something that may have overlapped with some other blended object you drew. Thus, the above becomes:
Render an object that reads from image, blends in the shader, and writes to image.
Call glTextureBarrier, if the next object overlaps with the first.
Render an object that reads from image, blends in the shader, and writes to image.
Call glTextureBarrier, if the next object overlaps with the first.
Repeat as needed for each blended object.

This must be done if you wish to read from any neighbouring texels. The whole primitive must be processed before the next one can start processing.

> The third thing this functionality changes is that you are permitted to perform a single read/modify/write operation between a texel fetch and a framebuffer image under the following conditions:
--  Each fragment shader reads only from a single texel from the image being written to. Specifically, the texel that that particular fragment shader will write to. This is easily done via texelFetch(sampler, ivec2(gl_FragCoord.xy), 0), though if the Viewport is adjusted from the 0,0 origin, you may need to bias the value by the viewport.
-- The read/modify/write happens only once between barrier calls. That is, you only perform one read/modify/write for each texel in the texture between calls to glTextureBarrier or other operations that ensure the visibility of writes. Note that even writing a value without reading counts, so you need a barrier before you start if you've already rendered to the image.

If you're interested in just the texel the fragment shader will overwrite, then the result is defined. Several primitives can be processed together, because only one texel is locked by the hardware.

This is just my understanding though, I have not worked with the extension.

**Comment by gonetz** [comment]
> - The read/modify/write happens only once between barrier calls. That is, you only perform one read/modify/write for each texel in the texture between calls to glTextureBarrier or other operations that ensure the visibility of writes.

Suppose, we draw two overlapping triangles with one glDrawArrays. 
Suppose P is a pixel covered by both triangles. The first triangle reads P.Z from the depth buffer and writes new P.Z back to the buffer. As I understand the cited sentence, it is one read/modify/write for that particular texel in the depth buffer texture. We need to call glTextureBarrier to ensure the visibility of that write to the second triangle, which also needs to modify P.Z
It also is just my understanding. From one side, your understanding of situation is too good to be true :)
From the other side, if I am correct the practical application of that extension is very limited.

**Comment by standard-two-simplex** [comment]
I have been reading more on the internet and I think you are right. So the extension only allows to do depth compare/blending by issues barrier calls, which is slow. That's a pity.

**Comment by standard-two-simplex** [comment]
I will close this, since it seems the extension doesn't work as I described in the first post.
